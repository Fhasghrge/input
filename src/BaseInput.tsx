import type { FC } from 'react';
import React, { cloneElement } from 'react';
import classNames from 'classnames';
import getInputCls from './utils/getInputCls';
import type { BaseInputProps } from './interface';
import { hasPrefixSuffix } from './utils/commonUtils';

const BaseInput: FC<BaseInputProps> = (props) => {
  const {
    inputElement,
    prefixCls,
    prefix,
    suffix,
    className,
    style,
    affixWrapperClassName,
    direction,
    disabled,
    readOnly,
    focused,
    allowClear,
    clearIcon,
    value,
    handleReset,
  } = props;

  // Clear Icon
  const getClearIcon = () => {
    if (!allowClear) {
      return null;
    }
    const needClear = !disabled && !readOnly && value;
    const clearIconCls = `${prefixCls}-clear-icon`;
    return (
      <span
        onClick={handleReset}
        // Do not trigger onBlur when clear input
        // https://github.com/ant-design/ant-design/issues/31200
        onMouseDown={(e) => e.preventDefault()}
        className={classNames(clearIconCls, {
          [`${clearIconCls}-hidden`]: !needClear,
          [`${clearIconCls}-has-suffix`]: !!suffix,
        })}
        role="button"
      >
        {clearIcon}
      </span>
    );
  };

  if (hasPrefixSuffix(props)) {
    const affixWrapperPrefixCls = `${prefixCls}-affix-wrapper`;
    const affixWrapperCls = classNames(
      affixWrapperPrefixCls,
      {
        [`${affixWrapperPrefixCls}-disabled`]: disabled,
        [`${affixWrapperPrefixCls}-focused`]: focused,
        [`${affixWrapperPrefixCls}-readonly`]: readOnly,
        [`${affixWrapperPrefixCls}-rtl`]: direction === 'rtl',
        [`${affixWrapperPrefixCls}-input-with-clear-btn`]:
          suffix && allowClear && value,
      },
      affixWrapperClassName,
      className,
    );

    const suffixNode = (suffix || allowClear) && (
      <span className={`${prefixCls}-suffix`}>
        {getClearIcon()}
        {suffix}
      </span>
    );

    return (
      <span className={affixWrapperCls} style={style}>
        {prefix && <span className={`${prefixCls}-prefix`}>{prefix}</span>}
        {cloneElement(inputElement, {
          style: null,
          value,
          className: getInputCls(prefixCls, disabled),
        })}
        {suffixNode}
      </span>
    );
  }

  return cloneElement(inputElement, {
    value,
  });
};

export default BaseInput;
