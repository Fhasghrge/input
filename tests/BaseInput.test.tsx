import type { ChangeEvent, FC } from 'react';
import React, { useState } from 'react';
import BaseInput from 'rc-input';
import { fireEvent, render, screen } from '@testing-library/react';

describe('BaseInput', () => {
  it('prefix and suffix should work', () => {
    const { container } = render(
      <div>
        <BaseInput
          prefixCls="rc-input"
          inputElement={<input />}
          prefix="prefix"
        />
        <br />
        <br />
        <BaseInput
          prefixCls="rc-input"
          inputElement={<input />}
          suffix="suffix"
        />
      </div>,
    );

    expect(container).toMatchSnapshot();
  });

  it('allowClear should work', () => {
    const onChange = jest.fn();

    const Demo: FC = () => {
      const [value, setValue] = useState<string>('');

      const handleReset = () => {
        setValue('');
      };

      const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
        onChange();
        setValue(e.target.value);
      };

      return (
        <BaseInput
          prefixCls="rc-input"
          allowClear
          clearIcon="âœ–"
          inputElement={<input onChange={handleChange} />}
          value={value}
          handleReset={handleReset}
        />
      );
    };

    const { container } = render(<Demo />);

    const inputEl = container.querySelector('input');
    fireEvent.change(inputEl!, { target: { value: 'some text' } });
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(inputEl!.value).toBe('some text');

    const clearIcon = container.querySelector('.rc-input-clear-icon');
    fireEvent.click(clearIcon!);
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(inputEl!.value).toBe('');
  });
});
