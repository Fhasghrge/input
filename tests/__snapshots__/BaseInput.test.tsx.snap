// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseInput prefix and suffix should work 1`] = `
<div>
  <div>
    <span
      class="rc-input-affix-wrapper"
    >
      <span
        class="rc-input-prefix"
      >
        prefix
      </span>
      <input
        class="rc-input"
        value=""
      />
    </span>
    <br />
    <br />
    <span
      class="rc-input-affix-wrapper"
    >
      <input
        class="rc-input"
        value=""
      />
      <span
        class="rc-input-suffix"
      >
        suffix
      </span>
    </span>
  </div>
</div>
`;
